### PR TITLE
feat: users/guimatheus92/v0.5.0

### PR DESCRIPTION
**Pull request by guimatheus92**
I have added a couple of features and improved the documentation.

What I have added:

- Plugin will not open until user has opened a project on SH3D, that way we can prevent errors and streamlining the user workflow:

image

- User can now choose betwen `HIGH` and `LOW` quality before exporting his images:
image

- The`Title` key-value from `.yaml` file can now be filled by user in SH3D in the `description` field in user project. That way, everytime a new file is generated, the data will be there saved in his project:

SH3D:

image

YAML:

``` yaml
  - type: state-icon
    entity: switch.kitchen_4
    title: Light Kitchen
    style:
      top: 70.37%
      left: 65.07%
      border-radius: 50%
      text-align: center
      background-color: rgba(255, 255, 255, 0.3)
    tap_action:
      action: toggle
```

- Since I am using Windows, the Makefile didn't work for me. So I created a PowerShell file following the same logic (great solution btw) that do the same work. 

**For Windows only**: I wasn't able to download the SweetHome3D `.jar` file using this method, so I have commented that line. For now, I suggest to download manually and place under a folder called `dl`

### CODE REVIEW

**Comment by guimatheus92** [comment]
@shmuelzon I don't if is possible to Pull Request my tag and my release as well, I wasn't able to do that. Or, if I should drag the file in here manually.

File: README.md [hunk 1]
```diff
@@ -8,6 +8,31 @@ light, sensors and cameras.
...
+- **Complete Renders Mode**: Renders images for all possible light combinations and rooms.
+- **YAML Configuration**: Generate a YAML file with the picture-elements structure for easy integration with Home Assistant.
+- **Configurations**:
+  - Group detected lights by room.
+  - Adjust output resolution (Width / Height).
+  - Adjust output quality (High / Low).
+  - Progress bar indicates rendering status.
+  - Create state icons for multiple entities from Home Assistant based on their names
+  - Lights is detected alone or within groups under SH3D
+  - Use existing exported renders to process quicker than the first time
```
Comment (shmuelzon) [inline README.md:24]:
```suggestion
  - Caches previously rendered images to save time when generating the next floorplan
```

File: README.md [hunk 2]
```diff
@@ -8,6 +8,31 @@ light, sensors and cameras.
...
+
+# How to activate plugin:
```
Comment (shmuelzon) [inline README.md:26]:
```suggestion
# How To Use The Plugin
```

File: README.md [hunk 3]
```diff
@@ -8,6 +8,31 @@ light, sensors and cameras.
...
+1. Download the latest release in `.jar` and copy to Sweet Home 3D "plugins" folder (see notes below) or download the latest release in `.sh3p` and execute that file.
```
Comment (shmuelzon) [inline README.md:27]: There are no `jar` files in the releases, only an `sh3p` file

File: README.md [hunk 4]
```diff
@@ -8,6 +8,31 @@ light, sensors and cameras.
...
+
+> 	If double-clicking on a `.sh3p` file doesn't launch Sweet Home 3D (most chances under Linux), you can also install a plug-in with the following command in a Terminal window (where SweetHome3D is the name of the executable file provided with Sweet Home 3D installers): `/path/to/SweetHome3D /path/to/plugin.sh3p`
```
Comment (shmuelzon) [inline README.md:29]: Please try to limit each line here (and in this file in general) to 80 characters

File: README.md [hunk 5]
```diff
@@ -8,6 +8,31 @@ light, sensors and cameras.
...
+
+2. Give the lights in Sweet Home 3D, the same name as in Home Assistant entities, for example: light.living_room. That way you will automate the name generation as well.
+3. Go to `Tools` in Sweet Home 3D -> Home Assistant Lightning Export
```
Comment (shmuelzon) [inline README.md:32]:
```suggestion
3. Go to `Tools` in Sweet Home 3D -> Home Assistant Floor Plan
```

File: README.md [hunk 6]
```diff
@@ -8,6 +8,31 @@ light, sensors and cameras.
...
+4. Move the generated images from Path to `/config/www/planes`
```
Comment (shmuelzon) [inline README.md:33]:
```suggestion
4. Copy the `floorplan` directory with all the images in it to your Home Assistant configuration directory `/config/www/`
```

File: README.md [hunk 7]
```diff
@@ -8,6 +8,31 @@ light, sensors and cameras.
...
+5. Create a card of type `picture-elements` and paste the generated `.yaml` from Path.
```
Comment (shmuelzon) [inline README.md:34]:
```suggestion
5. Create a card of type `picture-elements` and paste the contents of the generated `floorplan.yaml`  
```

File: README.md (hunk 2 above)
Comment (shmuelzon) [inline README.md:26]: Why do we need this entire section? It already exists under `Usage`

File: README.md [hunk 8]
```diff
@@ -92,8 +128,31 @@ When using the "Room overlay" light mixing mode, it's also suggested to:
   * It can later be converted to transparent using an image editor
 * Close all the doors between individually lighted rooms
 
+## Frequently Asked Questions
+
+1. **Where should I copy the generated files and what should I copy?**  
+  After the process is complete, copy the floorplan folder and floorplan.yaml to your Home Assistant path, specifically to: \config\www.
```
Comment (shmuelzon) [inline README.md:134]:
```suggestion
  After the process is complete, copy the floorplan folder and floorplan.yaml to your Home Assistant path, specifically to: /config/www.
```

File: README.md [hunk 9]
```diff
@@ -92,8 +128,31 @@ When using the "Room overlay" light mixing mode, it's also suggested to:
...
+
+2. **How do I select the desired perspective for rendering?**  
+  Before activating the plugin, set the SH3D project to the specific 3D point of view that you want to be rendered.
+
+3. **How do I change the rendering settings?**  
+  Prior to activating the plugin, go to “Create photo…” in the SH3D project and adjust the settings there. You do not need to render or save anything; simply make the changes and close the dialog.
```
Comment (shmuelzon) [inline README.md:140]:
```suggestion
  Prior to activating the plugin, go to "Create photo…" in the SH3D project and adjust the settings there. You do not need to render or save anything; simply make the changes and close the dialog.
```

File: README.md [hunk 10]
```diff
@@ -107,3 +166,10 @@ When using the "Room overlay" light mixing mode, it's also suggested to:
 - [ ] Allow overriding state-icons/labels positions, and save persistently
 - [ ] Allow defining, per entity, if it should be an icon or label, and save
       that persistently
+
+## Contribute
```
Comment (shmuelzon) [inline README.md:170]: I think this section should go in a separate `CONTRIBUTING.md` file

File: README.md [hunk 11]
```diff
@@ -107,3 +166,10 @@ When using the "Room overlay" light mixing mode, it's also suggested to:
...
+
+If you are using a UNIX system, just run the `Makefile` and it will generate the necessary build to run this plugin.
```
Comment (shmuelzon) [inline README.md:172]:
```suggestion
If you are using a UNIX system, just run `make` to build the plugin. `make test` will install the plugin and start SH3D.
```

File: build.ps1 [hunk 12]
```diff
@@ -0,0 +1,105 @@
+param (
```
Comment (shmuelzon) [inline build.ps1:1]: I'm sorry, but I would rather not maintain two different build scripts.
I'm also on Windows and just use WSL to build the plugin

File: src/com/shmuelzon/HomeAssistantFloorPlan/Controller.java [hunk 13]
```diff
@@ -77,8 +80,32 @@ public Controller(Home home) {
         lightsNames = new ArrayList<String>(lights.keySet());
         lightsPower = getLightsPower(lights);
         homeAssistantEntities = getHomeAssistantEntities();
+        this.quality = Quality.HIGH;
+    }
+
+    public Home getHome() {
```
Comment (shmuelzon) [inline src/com/shmuelzon/HomeAssistantFloorPlan/Controller.java:86]: I'd rather not expose `Home` directly to the UI. If you really want to indicate the project is empty, create a function specifically for that

File: src/com/shmuelzon/HomeAssistantFloorPlan/Controller.java [hunk 14]
```diff
@@ -492,18 +520,18 @@ private Point2d getFurniture2dLocation(HomePieceOfFurniture piece) {
         return new Point2d((objectPosition.x * 0.5 + 0.5) * renderWidth, (objectPosition.y * 0.5 + 0.5) * renderHeight);
     }
 
-    private String generateStateIconYaml(String name, Point2d position, String type, String action) {
+    private String generateStateIconYaml(String name, Point2d position, String type, String action, String title) {
         String yaml = String.format(Locale.US,
             "  - type: state-%s\n" +
             "    entity: %s\n" +
-            "    title: null\n" +
+            "    title: %s\n" +
```
Comment (shmuelzon) [inline src/com/shmuelzon/HomeAssistantFloorPlan/Controller.java:527]: If the description is empty, what will appear here (and in the HA frontend)? Will the behavior remain the same?

File: src/com/shmuelzon/HomeAssistantFloorPlan/Controller.java [hunk 15]
```diff
@@ -532,7 +560,7 @@ private List<StateIcon> generateLightsStateIcons() {
                 lightsCenter.add(getFurniture2dLocation(light));
             lightsCenter.scale(1.0 / lightsList.size());
 
-            stateIcons.add(new StateIcon(lightsList.get(0).getName(), lightsCenter, "icon", "toggle"));
+            stateIcons.add(new StateIcon(lightsList.get(0).getName(), lightsCenter, "icon", "toggle", lightsList.get(0).getDescription() != null ? lightsList.get(0).getDescription() : lightsList.get(0).getName()));
```
Comment (shmuelzon) [inline src/com/shmuelzon/HomeAssistantFloorPlan/Controller.java:563]: I think I prefer the title to remain null and not have a tooltip with the name of the entity

File: src/com/shmuelzon/HomeAssistantFloorPlan/Controller.java [hunk 16]
```diff
@@ -558,7 +586,8 @@ private List<StateIcon> generateSensorsStateIcons() {
 
         for (HomePieceOfFurniture piece : homeAssistantEntities) {
             Point2d location = getFurniture2dLocation(piece);
-            stateIcons.add(new StateIcon(piece.getName(), location, piece.getName().startsWith("sensor.") ? "label" : "icon", isHomeAssistantEntityActionable(piece.getName()) ? "toggle" : null));
+            String title = piece.getDescription() != null ? piece.getDescription() : piece.getName();
```
Comment (shmuelzon) [inline src/com/shmuelzon/HomeAssistantFloorPlan/Controller.java:589]: Same

File: src/com/shmuelzon/HomeAssistantFloorPlan/Panel.java [hunk 17]
```diff
@@ -1,17 +1,7 @@
...
-import java.awt.GridBagLayout;
-import java.awt.Insets;
-import java.awt.event.ItemEvent;
-import java.awt.event.ItemListener;
-import java.awt.Window;
-import java.awt.event.ActionEvent;
-import java.awt.event.ActionListener;
-import java.awt.event.WindowAdapter;
-import java.awt.event.WindowEvent;
+import java.awt.*;
```
Comment (shmuelzon) [inline src/com/shmuelzon/HomeAssistantFloorPlan/Panel.java:3]: I prefer explicit imports and not wildcards

**Review by shmuelzon — CHANGES_REQUESTED** [review]

**Comment by shmuelzon** [comment]
Thank you for the pull request!
Would it be possible to split each feature/change to a separate pull request?
That way it'll show up properly in the change log.

Thanks!

**Comment by guimatheus92** [comment]
> Thank you for the pull request!
> 
> Would it be possible to split each feature/change to a separate pull request?
> 
> That way it'll show up properly in the change log.
> 
> 
> 
> Thanks!

Sure, I will take some time do undo all I did and create a PR for each but I will do. There a couple of comments there that I don't know how to solve, so I might open an issue to ask help/ideas so I can implement.

**Pull request closed without merging**